### PR TITLE
refactor path model

### PR DESCRIPTION
**Pull request by tomerd**
motivation: deprecate TSC, move to swift-system

changes:
* create AbsolutePath and RelativePath in SwiftPM, at this point just a wrapper for the TSC version. In next PR we will start refactoring them
* remove all direct imports of TSBasic and replace them with data type specific imports to help migration
* reduce sue of TSCTestSupport, stop exporting it from SPMTestSupport
* update all call sites and tests

### CODE REVIEW

**Comment by tomerd** [comment]
https://github.com/apple/swift-tools-support-core/pull/418

@swift-ci smoke test

File: Sources/Basics/FileSystem/FileSystem+Extensions.swift [hunk 1]
```diff
@@ -10,25 +10,209 @@
...
+import protocol TSCBasic.FileSystem
+import var TSCBasic.localFileSystem
+import protocol TSCBasic.WritableByteStream
+
+public typealias FileSystem = TSCBasic.FileSystem
+public var localFileSystem = TSCBasic.localFileSystem
+
+// MARK: - Custom path
+
+extension FileSystem {
```
Comment (tomerd) [inline Sources/Basics/FileSystem/FileSystem+Extensions.swift:31]: this is an adapter so FS work with the "new" types

File: Sources/Basics/FileSystem/RelativePath.swift [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+/// - Collapsing `..` path components that aren't at the beginning
+/// - Removing extraneous `.` path components
+/// - Removing any trailing path separator
+/// - Removing any redundant path separators
+/// - Replacing a completely empty path with a `.`
+///
+/// This string manipulation may change the meaning of a path if any of the
+/// path components are symbolic links on disk.  However, the file system is
+/// never accessed in any way when initializing a RelativePath.
+public struct RelativePath: Hashable, Sendable {
```
Comment (tomerd) [inline Sources/Basics/FileSystem/RelativePath.swift:33]: this is the key change

File: Sources/Basics/FileSystem/TSCFreeFunctionsAdapter.swift [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
```
Comment (tomerd) [inline Sources/Basics/FileSystem/TSCFreeFunctionsAdapter.swift:11]: this is a shim between utility functions in TSC and the "new" types

File: Sources/Basics/FileSystem/VirtualFileSystem.swift [hunk 4]
```diff
@@ -119,18 +133,31 @@ private extension FileSystem {
 public class VirtualFileSystem: FileSystem {
```
Comment (tomerd) [inline Sources/Basics/FileSystem/VirtualFileSystem.swift:133]: concrete FS implementation use the TSC types

**Comment by tomerd** [comment]
https://github.com/apple/swift-tools-support-core/pull/418

@swift-ci smoke test

**Comment by tomerd** [comment]
smoke tests expected to fail since they do not support cross repo testing

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test windows

**Comment by tomerd** [comment]
cc @abertelrud @compnerd this is the first step towards cleaning this up

**Comment by tomerd** [comment]
https://github.com/apple/sourcekit-lsp/pull/741

@swift-ci smoke test

**Comment by tomerd** [comment]
https://github.com/apple/sourcekit-lsp/pull/741

@swift-ci smoke test

File: Sources/Basics/FileSystem/AbsolutePath.swift [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+    /// Extension of the give path's basename. This follow same rules as
+    /// suffix except that it doesn't include leading `.` character.
+    public var `extension`: String? {
+        self.underlying.extension
+    }
+
+    /// Unlike ``AbsolutePath//extension``, this property returns all characters after the first `.` character in a
+    /// filename. If no dot character is present in the filename or first dot is the last character, `nil` is returned.
+    public var allExtensions: [String]? {
```
Comment (neonichu) [inline Sources/Basics/FileSystem/AbsolutePath.swift:117]: I guess this was a `Basics`-provided extension before? Might be worth keeping that separation?

File: Sources/Basics/FileSystem/AbsolutePath.swift [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+    /// to be a valid path component (i.e., it cannot be empty, contain a path
+    /// separator, or be a pseudo-path like '.' or '..').
+    public func appending(_ components: String...) -> AbsolutePath {
+        self.appending(components: components)
+    }
+
+    /// Returns the absolute path with additional extension appended.
+    ///
+    public func appending(extension: String) -> AbsolutePath {
+        guard !self.isRoot else { return self }
```
Comment (neonichu) [inline Sources/Basics/FileSystem/AbsolutePath.swift:223]: same as above

File: Sources/Basics/Netrc.swift [hunk 7]
```diff
@@ -74,7 +76,7 @@ public struct Netrc {
     }
 }
 
-public struct NetrcParser {
+public enum NetrcParser {
```
Comment (neonichu) [inline Sources/Basics/Netrc.swift:79]: 🤔

File: Sources/Basics/SwiftVersion.swift [hunk 8]
```diff
@@ -23,11 +23,11 @@ public struct SwiftVersion {
     public var buildIdentifier: String?
 
     /// The major component of the version number.
-    public var major: Int { return self.version.major }
```
Comment (neonichu) [inline Sources/Basics/SwiftVersion.swift:26]: 😭

File: Sources/Basics/Netrc.swift (hunk 7 above)
Comment (tomerd) [reply Sources/Basics/Netrc.swift:79]: swift format made me do it

Comment (neonichu) [reply Sources/Basics/Netrc.swift:79]: That's pretty weird?

**Comment by tomerd** [comment]
https://github.com/apple/sourcekit-lsp/pull/741

@swift-ci smoke test

**Comment by tomerd** [comment]
https://github.com/apple/sourcekit-lsp/pull/741

@swift-ci smoke test windows

**Comment by tomerd** [comment]
@compnerd does the windows build failure make sense you? looks like a compiler crash

**Comment by compnerd** [comment]
It is an assertion failure:
```
Assertion failed: Ptr && "Cannot dereference a null Type!", file C:\Users\swift-ci\jenkins\workspace\swiftpm-PR-windows\swift\include\swift/AST/Type.h, line 249
```

Looks like a swift compiler issue.

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by tomerd** [comment]
@compnerd its only crashing in windows tho, is this something you can help with?

**Comment by compnerd** [comment]
I think that @xymus would be better able to help than I would.  However, the last run seems to be failing differently:

```
C:\Users\swift-ci\jenkins\workspace\swiftpm-PR-windows\sourcekit-lsp\Sources\SKSwiftPMWorkspace\SwiftPMWorkspace.swift:108:59: error: cannot convert value of type 'TSCBasic.AbsolutePath' to expected argument type 'Basics.AbsolutePath'

    var location = try Workspace.Location(forRootPackage: packageRoot, fileSystem: fileSystem)

                                                          ^

C:\Users\swift-ci\jenkins\workspace\swiftpm-PR-windows\sourcekit-lsp\Sources\SKSwiftPMWorkspace\SwiftPMWorkspace.swift:179:51: error: cannot convert value of type 'TSCBasic.AbsolutePath' to expected element type 'Basics.AbsolutePath'

      rootInput: PackageGraphRootInput(packages: [packageRoot]),

                                                  ^

C:\Users\swift-ci\jenkins\workspace\swiftpm-PR-windows\sourcekit-lsp\Sources\SKSwiftPMWorkspace\SwiftPMWorkspace.swift:238:28: error: cannot convert return expression of type 'Basics.AbsolutePath' to return type 'TSCBasic.AbsolutePath'

    return buildParameters.buildPath

           ~~~~~~~~~~~~~~~~^~~~~~~~~

C:\Users\swift-ci\jenkins\workspace\swiftpm-PR-windows\sourcekit-lsp\Sources\SKSwiftPMWorkspace\SwiftPMWorkspace.swift:242:51: error: cannot convert return expression of type 'AbsolutePath' to return type 'AbsolutePath?'

    return buildParameters.indexStoreMode == .off ? nil : buildParameters.indexStore

           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

C:\Users\swift-ci\jenkins\workspace\swiftpm-PR-windows\sourcekit-lsp\Sources\SKSwiftPMWorkspace\SwiftPMWorkspace.swift:360:19: error: cannot convert value of type 'TSCBasic.AbsolutePath' to expected argument type 'Basics.AbsolutePath'

        filePath: path, 

                  ^

C:\Users\swift-ci\jenkins\workspace\swiftpm-PR-windows\sourcekit-lsp\Sources\SKSwiftPMWorkspace\SwiftPMWorkspace.swift:422:55: error: binary operator '==' cannot be applied to operands of type 'TSCBasic.AbsolutePath' and 'Basics.AbsolutePath'

      for package in packageGraph.packages where path == package.manifest.path {

                                                 ~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~

C:\Users\swift-ci\jenkins\workspace\swiftpm-PR-windows\sourcekit-lsp\Sources\SKSwiftPMWorkspace\SwiftPMWorkspace.swift:506:70: error: binary operator '==' cannot be applied to operands of type 'Basics.AbsolutePath' and 'TSCBasic.AbsolutePath'

    let compilePath = try td.compilePaths().first(where: { $0.source == nativePath })

                                                           ~~~~~~~~~ ^  ~~~~~~~~~~
```

**Comment by tomerd** [comment]
@compnerd I think its because @MaxDesiatov forgot to include the LSP PR

**Comment by tomerd** [comment]
https://github.com/apple/sourcekit-lsp/pull/741

@swift-ci smoke test windows

**Comment by tomerd** [comment]
https://github.com/apple/sourcekit-lsp/pull/741

@swift-ci smoke test windows

**Comment by tomerd** [comment]
https://github.com/apple/sourcekit-lsp/pull/741

@swift-ci smoke test

**Comment by tomerd** [comment]
https://github.com/apple/sourcekit-lsp/pull/741

@swift-ci smoke test windows

**Pull request merged**